### ISSUE TITLE
Classification tree outputs learner when given new data

### ISSUE THREAD

**Issue by janezd** [biolab/orange3#1044]
Construct a workflow in which you have a File widget and Classification tree widget, both connected to Test learners (note that File is *not* connected to Classification tree). Test Learners runs cross validation and shows the results.

Now connect File to Tree. The tree widget should output only the learner, but it also outputs the tree learner again (although it does not change when the data is changed), causing the Test Learners to unnecessarily recompute the cross validation.

**Comment by s-alexey** [comment]
I've suggested my solution. 

But it isn't only `OWClassificationTree` problem. I think it would be better to add some logic into `OWProvidesLearner` class. 

something like:

```python
    def apply(self):
        self.update_learner()
        self.update_model()
        
    def update_learner(self):
        raise NotImplementedError()
    
    def update_model(self):
        raise NotImplementedError()
```

**Comment by kernc** [comment]
Code consolidation and refactoring are an integral part of a healthy development process. As you say, the more one can solve problems cleanly, the better.

**Comment by s-alexey** [comment]
Ok, I can perform `OWProvidersLearner` improvement. Here is a drawft. I would like to know you suggestions.
```python 
class OWProvidesLearner(_OWAcceptsPreprocessor):
    """
    Base class for all classification / regression learner-providing widgets
    that extend it.
    """

    LEARNER = None
    OUTPUT_MODEL_NAME = None
    OUTPUT_MODEL_CLASS = None

    inputs = [("Data", Table, "set_data")] + _OWAcceptsPreprocessor.inputs

    @property
    def outputs(self):
        return [
            ("Learner", self.LEARNER),
            (self.OUTPUT_MODEL_NAME, self.OUTPUT_MODEL_CLASS)
        ]

    model_name = Setting(OUTPUT_MODEL_NAME)

    def __init__(self):
        super().__init__()

        self.data = None
        self.learner = None
        self.model = None
        # here can be gui configuration (e.g. gui.lineEdit for model_name)
        # we can add method get_setting_box() 
        # 
        # gui.lineEdit
        # box = get_setting_box()
        # apply and report buttons 

    @check_sql_input
    def set_data(self, data):
        """Set the input train data set."""
        self.error(0)
        self.data = data
        if data is not None and data.domain.class_var is None:
            self.error(0, "Data has no target variable")
            self.data = None
        self.update_model()

    def apply(self):
        self.update_learner()
        self.update_model()

    def update_learner(self):
        # ...
        # self.learner.name = self.model_name
        # self.send("Learner", self.learner)
        raise NotImplementedError()

    def update_model(self):
        self.model = None
        if self.data is not None:
            # 0 or 1?
            self.error(1)
            if not self.learner.check_learner_adequacy(self.data.domain):
                self.error(1, self.learner.learner_adequacy_err_msg)
            else:
                self.model = self.learner(self.data)
                self.model.name = self.model_name
                self.model.instances = self.data

        self.send(self.OUTPUT_MODEL_NAME, self.model)

    def send_report(self):
        raise NotImplementedError()
```

I have tried it with TreeClassification and it works fine.

**Comment by kernc** [comment]
:+1: 

`outputs` property would have to be writable, see https://github.com/biolab/orange3/blob/506c87c8b42a7842e47072c1dbe5207459eb4dfb/Orange/widgets/widget.py#L39.

Indeed, there's a lot that could be beneficially consolidated. If you submit this as a "[WIP]" pull request, we can comment on lines individually.

**Comment by astaric** [comment]
Fixed in #1072
